crates/sui-core/src/unit_tests/authority_tests.rs

```diff
@@ -953,6 +1050,57 @@ async fn test_dev_inspect_on_validator() {
     assert!(result.is_err())
 }
 
+#[tokio::test]
+async fn test_dry_run_transaction_types() {
+    let (valid_transaction_kinds, invalid_transaction_kinds) = transaction_types();
+
+    let (sender, sender_key): (_, AccountKeyPair) = get_key_pair();
+    let gas_object_id = ObjectID::random();
+    let (_validator, fullnode, _) =
+        init_state_with_ids_and_object_basics_with_fullnode(vec![(sender, gas_object_id)]).await;
+    let gas_object = fullnode.get_object(&gas_object_id).await.unwrap().unwrap();
+
+    for valid_tx in valid_transaction_kinds.into_iter() {
+        let rgp = fullnode.reference_gas_price_for_testing().unwrap();
+        let tx_data = TransactionData::new(
+            valid_tx,
+            sender,
+            gas_object.compute_object_reference(),
+            rgp * TEST_ONLY_GAS_UNIT_FOR_OBJECT_BASICS,
+            rgp,
+        );
+        let tx = to_sender_signed_transaction(tx_data.clone(), &sender_key);
+        let digest = tx.digest();
+        let res = fullnode
+            .dry_exec_transaction(tx_data, *digest)
+            .await
+            .unwrap()
+            .0;
+        assert!(res.effects.status().is_ok());
+    }
+
+    for invalid_tx in invalid_transaction_kinds.into_iter() {
+        assert!(invalid_tx.is_system_tx());
+        let rgp = fullnode.reference_gas_price_for_testing().unwrap();
+        let tx_data = TransactionData::new(
+            invalid_tx,
+            sender,
+            gas_object.compute_object_reference(),
+            rgp * TEST_ONLY_GAS_UNIT_FOR_OBJECT_BASICS,
+            rgp,
+        );
+        let tx = to_sender_signed_transaction(tx_data.clone(), &sender_key);
+        let digest = tx.digest();
+        let res = fullnode.dry_exec_transaction(tx_data, *digest).await;
+
+        assert!(res.is_err());
+        assert!(matches!(
+            res.unwrap_err(),
+            SuiError::UnsupportedFeatureError { .. }
```

Comment:
I could see it going one way or the other (IMO I don't believe the current error is wrong, but can see the argument to change it as well). I'd lean towards keeping it as-is just because this lessens possible churn on clients, but we can change the error here if folks have strong feelings on it.